src/grt/test/grt_aux.py

```diff
@@ -0,0 +1,157 @@
+import utl
+from enum import Enum
+
+
+def get_layer_idx(design, layer_name):
+    """ Given a layer name and a technology, return the layer index"""
+    mt = design.getTech().getDB().getTech()
+    if mt == None:
+        utl.error(utl.GRT, 500, "No technology has been read.")
+    tlayer = mt.findLayer(layer_name)
+    if tlayer == None:
+        utl.error(utl.GRT, 501, f"Layer {layer_name} not found.")
+    idx = tlayer.getRoutingLevel()
+    return idx
+
+
+def get_layer_range(design, layer_range):
+    l1, l2 = layer_range.split("-")
+    idx1 = get_layer_idx(design, l1)
+    idx2 = get_layer_idx(design, l2)
+    if idx1 > idx2:
+        utl.error(utl.GRT, 502, "Input format to define layer range is min-max.")
+    return idx1, idx2
+
+
+# This also assumes that routing layer indicies are contiguous
+def check_routing_layer_by_index(design, layer_idx):
+    """Check that the layer index 'layer_idx' is in bounds of the defined layers"""
+    rtech = design.getTech().getDB().getTech()
+    if rtech == None:
+        utl.error(utl.GRT, 503, "No technology has been read.")
+    if layer_idx < 0:
+        utl.error(utl.GRT, 504, "Layer index must be non negative.")
+
+    # this is actually the max index, where they start at 0, not the total
+    # number of routing layers
+    max_routing_layer = rtech.getRoutingLayerCount()
+    tech_layer = rtech.findRoutingLayer(layer_idx)
+    
+    # NOTE: the tcl version uses 1 instead of 0, which is incorrect.
+    min_tech_layer = rtech.findRoutingLayer(0)
+    max_tech_layer = rtech.findRoutingLayer(max_routing_layer)
+
+    if layer_idx > max_routing_layer:
+        utl.error(utl.GRT, 505, f"Layer {tech_layer.getConstName()} is greater "
+                                f"than the max routing layer "
+                                f"({max_tech_layer.getConstName()}).")
+        
+    if layer_idx < 0:
+        utl.error(utl.GRT, 506, f"Layer {tech_layer.getConstName()} is less "
+                                f"than the min routing layer "
+                                f"({min_tech_layer.getConstName()}).")
+
+
+def set_global_routing_layer_adjustment(design, layer, adj):
+    """In design "design", set the global layer adjustment for "layer".
+       layer can be a named layer or range specified as layer1-layer2,
+       where layer1 < layer2. The adjustment "adj" must be > 0.0"""
+    gr = design.getGlobalRouter()
+    
+    if adj <= 0.0:
+        utl.error(utl.GRT, 507, "Layer adjustment must be positive")
```

Comment:
should be checked in c++

Reply:
So effectively, that means moving much of grt_aux.py into the C++ api. You comment below that "In general it feels like there is too much support code in the unit tests..." I can understand that but the "support code" for the tcl interface lies in both the GlobalRouter.i and the GlobalRouter.tcl files and totals about 800 lines (let's de-rate to 700 for boilerplate). The GlobalRouter-py.i file and grt_aux.py total about 200 loc, so that is still quite a step down. It will be better in C++ for sure, but it is a redesign of the api.